File: go/solution/0003.go
//
// Solution for Project Euler problem 3
// Copyright michimani All rights reserved.
//
// https://projecteuler.net/problem=3
//
package solution

import (
	"fmt"
)

// Solve0003 is function to solve the problem.
func Solve0003() {
	var answer int
	var primesMap map[int]int = map[int]int{}
	var isPrime bool = true

	var targetNumber int = 600851475143
	var targetNumberDiv int = targetNumber

	for num := 2; num < targetNumber; num++ {
		isPrime = true
		for _, val := range primesMap {
			if num%val == 0 {
				isPrime = false
				break
			}
		}

		if isPrime == true {
			primesMap[num] = num

			for targetNumberDiv%num == 0 {
				targetNumberDiv = targetNumberDiv / num
			}

			if targetNumberDiv <= 1 {
				answer = num
				break
			}
		}

	}

	fmt.Println(answer)
}

File: python/0003.py
#
# Solution for Project Euler problem 3
# Copyright michimani All rights reserved.
#
# https://projecteuler.net/problem=3
#


def compute():
    max_num = 600851475143
    quot = max_num
    primes = []
    ans = 0
    for prime_chk in range(2, max_num):
        is_prime = True
        for prime in primes:
            if prime_chk % prime == 0:
                is_prime = False
                break

        if is_prime is True:
            primes.append(prime_chk)
            while quot % prime_chk == 0:
                quot = int(quot / prime_chk)

            if quot <= 1:
                ans = prime_chk
                break

    return ans


if __name__ == "__main__":
    print(compute())
